### ISSUE TITLE
Meson ignores RCFLAGS in the windows resources compiler

### ISSUE THREAD

**Issue by ylatuya** [mesonbuild/meson#4736]
In a multilib gcc+mingw toolchain, the windows resources compiler `x86_64-w64-mingw32-windres` needs to have the `-F pe-i386` option set so it generates binaries for the correct architecture.
Normally these options are passed through RCFLAGS but mesonbuild/modules/windows.py seems to ingore them.

**Comment by jpakkane** [comment]
Those should be set via the cross file when cross compiling. Envvars are only for native compilation.

We don't seem to currently have support for this but you can set the rc compiler in the cross file as an array, so something like `['x86_64-w64-mingw32-windres', '-F', 'pe-i386']`.

**Comment by ylatuya** [comment]
I fixed it this way in our build system, by exporting WINDRES="x86_64-w64-mingw32-windres -F pe-i386" or RC="x86_64-w64-mingw32-windres -F pe-i386"  (we export both, I don't know which is one is picked), but I think meson should support RCFLAGS.

For example when compiling a c file, with a multilib toolchain I don't have to change anything in the cross file that sets the compiler to be   "x86_64-w64-mingw32-gcc -m32", I simply export CFLAGS="-m32", in the same way for windres, we should be able to export RCFLAGS="-F pe-i386". But these are problably just hacks since the build system should set those correctly knowing the host architecture.

If I understood correctly from your comments, the correct fix in meson should  be adding this option directly in the resources compiler implementation, depending on the output architecture, so that independently of if it's a multilib toolchain or not,  -F is always set to output architecture. I am correct?

**Comment by nirbheek** [comment]
Since you want `-m32` with every compiler invocation, you should set it as part of `WINDRES` (if compiling natively) and in the cross-file as `windres = ['x86_64-w64-mingw32-windres', '-m32']` if cross-compiling.

**Comment by nirbheek** [comment]
Meson can't know what sort of toolchain you have, so we can't set it automatically. We've been discouraging people from using env variables for the build system, and since you can always set `WINDRES` and `windres = ` in the cross file, I don't think there is a particular need for `RCFLAGS`.

**Comment by ylatuya** [comment]
I think there  seems to be an issue with native builds, cross build works now correctly with the cross file.
In native builds, when setting WINDRES="x86_64-w64-mingw32-windres -F pe-i386", windres.get_command() returns ["x86_64-w64-mingw32-windres -F pe-i386"] instead of ['x86_64-w64-mingw32-windres', '-F', 'pe-i386']. I believe it's not splitting correctly the values from the env var.

**Comment by ylatuya** [comment]
I have tried tried setting it in the native file withing the [binaries] section but it has the same problem.

**Comment by Ericson2314** [comment]
CC @dcbaker TODO when we do "structured: properties sections.

**Comment by lhmouse** [comment]
Actually I want to do this

```meson
if get_option('enable-debug-checks')
  add_project_arguments('-DMY_DEBUG_BUILD', language: [ 'c', 'rc' ])
endif
```

but at the moment `rc` isn't a valid language name. Would it make sense to add `rc` as a language so `import('windows').compile_resources(...)` can have implicit language-specific arguments, like above?

**Comment by Ericson2314** [comment]
So totally coincidentally from me commenting on this feature 7 years ago, I recently did https://github.com/mesonbuild/meson/pull/15438, which is somewhat analogous. Perhaps that can help?


@lhmouse Yes I agree, the right way to support `RCFLAGS` is to bump it to a language.

**Comment by Ericson2314** [comment]
@lhmouse I am gonna put up a draft untestd slop PR that does the refactor to make rc a real language not not extension module hacks.

**Comment by bonzini** [comment]
No, don't. Just add a built-in rc_args option and tie it to RCFLAGS.

**Comment by Ericson2314** [comment]
I mean, I think it is good to at least skim the new methodology.

I am not against small extension modules, but IMO this one already crossed the rubicom of trying to support multiple different compilers etc. At some point, it is better to use existing abstractions than continue hand-rolling a cheap knock-off of the compiler infra with custom targets.

**Comment by bonzini** [comment]
Possibly but adding the flags option doesn't hamper backwards compatibility when adding a resource language later. But it's a much smaller task, and the compiler might require extra infrastructure changes to support preprocessing. It's not simple.

**Comment by Ericson2314** [comment]
https://github.com/mesonbuild/meson/pull/15679 OK the PR is up. I think it the new code is very elegant --- much more so. However, I am worried about the treatment of manual `depends` and `depend_files`.

**Comment by eli-schwartz** [comment]
> [@lhmouse](https://github.com/lhmouse) I am gonna put up a draft untestd slop PR that does the refactor to make rc a real language not not extension module hacks.

~~Can you confirm whether your PR is LLM generated?~~ EDIT: sigh

**Comment by Ericson2314** [comment]
(I just saw this now.) I used the LLM as I do for all large rote refactors since last fall. And a very substantial part of what my work across all projects has always been is large refactors.

**Comment by lhmouse** [comment]
Off-topic but I have a strong prejudice about code that is generated by LLM. However it's fine to use LLM to do auxiliary code review, or to create testcases (but those need to be marked so).

**Comment by Ericson2314** [comment]
What I found pre LLM is that I would carefully set up a refactor, and then spend a huge amount of time mindlessly grinding through type errors to fix everything. (This is a large reason why I have always strongly preferred typed languages.) 

I was highly skeptical of the sort of "I don't care what the code looks like" people churning out all green no red PRs, just as I was highly skeptical of the human-written slop those people would do before LLMs. But once I realized that the LLMs were good enough to do the "mindlessly finish the refactor"' part, I was sold. I think it's a good way to get a lot of value out while giving the tool little to no creative control, keeping it on a very tight leash.

**Comment by lhmouse** [comment]
> What I found pre LLM is that I would carefully set up a refactor, and then spend a huge amount of time mindlessly grinding through type errors to fix everything. (This is a large reason why I have always strongly preferred typed languages.)
> 
> I was highly skeptical of the sort of "I don't care what the code looks like" people churning out all green no red PRs, just as I was highly skeptical of the human-written slop those people would do before LLMs. But once I realized that the LLMs were good enough to do the "mindlessly finish the refactor"' part, I was sold. I think it's a good way to get a lot of value out while giving the tool little to no creative control, keeping it on a very tight leash.

I can't agree with this. Sorry.

It's purely psychological reason: Using AI suggestion might be fine, but when you write something you have to think; it's like you watch other people write then you don't think at all. It's wrong.

**Comment by Ericson2314** [comment]
Reading doesn't force thinking like writing, but it doesn't preclude it either. I like reading code, and I think a lot when doing it.
